### ISSUE TITLE
Interactive Book does not return any results to the calling LMS

### ISSUE THREAD

**Issue by csidirop** [h5p/h5p-interactive-book#154]
The Interactive Book module does not return any results to the calling LMS (in our case [ILIAS](https://en.wikipedia.org/wiki/ILIAS)). 

Other H5P modules like [Blanks](https://github.com/h5p/h5p-blanks) or [Drag Text](https://github.com/h5p/h5p-drag-text) returning how many tasks were solved correctly (from all of them).

I'm sure other LMSs using the results too. Something like: 
 - Test1: 2/2 (if two out of two were correct answered)
 - Test2: 0/1 

I couldn't verify this, but I believe IB is missing a file `presave.js` (like [presave.js in drag-text](https://github.com/h5p/h5p-drag-text/blob/master/presave.js)) or [presave.jsin blanks](https://github.com/h5p/h5p-blanks/blob/master/presave.js). For example, it validates the content for the Blanks module, calculates the maximum score based on the number of blanks in the questions, and ensures that the required parameters are present before saving the content. If any problems are found, it throws an error, otherwise it validates the score and calls the completion callback.

Is that correct? Can we have that too? 

If yes, I could provide help or implement it myself if you don't have resources left for that kind of task.

**Comment by otacke** [comment]
The presave.js file is not a requirement. It's an optional addition (only used by Edlib to the best of my knowledge) that would allow to determine the number of maximum points before the content is rendered for the first time (if you desperately required to know the maximum possible score server side immediately after saving the content), but it has nothing to do with the xAPI data that is used for reporting.

Interactive Book only reports scores when the user hits the "Submit Report" button on the Summary page (just like Interactive Video only reports scores when the user submits them or like Course Presentation only submits them when the user enters the summary slide). That's more likely the cause.

If you don't see the "Submit Report" button on ILIAS, you should reach out to Studer&Raimann, the maintainers of the ILIAS module. They'd have to ensure that their H5P integration sets the "environment" variable `H5PIntegration.reportingIsEnabled` to `true`.

**Comment by csidirop** [comment]
Thanks for the quick clarification! 

I will take a closer look and report back.

**Comment by csidirop** [comment]
A little addendum:

If I force `H5PIntegration.reportingIsEnabled` to `true` somewhere after: https://github.com/h5p/h5p-interactive-book/blob/a3b1cdeb09881c90615fb715483e545ec2a983fd/src/scripts/app.js#L56-L63
I'm getting the submit button. Thats what it looks like for those who wonder:
![grafik](https://github.com/user-attachments/assets/444ed43c-187e-498f-8115-8920217d0f34)

After a bit of searching I found a forum post: [Interactive Book: Missing "Submit Report" Button](https://h5p.org/node/1026902) in which you gave the hint to:

> ... tick the option "Enable LRS dependent content types" in the plugin settings of H5P which will then set a flag for reporting being enabled (H5PIntegration.reportingIsEnabled as mentioned on the issue tracker). It's a non-intuitive procedure unfortunately.

Unfortunately that didn't changed anything in my tests (ILIAS r_7 & r_9). But added it for future reference. Is this variable even in use anymore? I couldn't find any trace of `H5PIntegration.reportingIsEnabled` or `reportingIsEnabled` other than here and at the [H5P Documentation Tool](https://github.com/h5p/h5p-documentation-tool/blob/d6eddeaa200dd490ebcace443d924994a35efe2a/documentation-tool.js#L32). (grep over the whole ILIAS dir)

**Comment by csidirop** [comment]
> you should reach out to Studer&Raimann, the maintainers of the ILIAS module. They'd have to ensure that their H5P integration sets the "environment" variable `H5PIntegration.reportingIsEnabled` to `true`.

[Ticket: SUPPORT-15210](https://jira.sr.solutions/servicedesk/customer/portal/27/SUPPORT-15210)

**Comment by 0xMurage** [comment]
I should leave this here for reference:

The `isScoringEnabled` and `isReportingEnabled` flags are not supported by the H5P core library. The core library's third parameter only sends the following in the third parameter, as seen here 
https://github.com/h5p/h5p-php-library/blob/fb5100a53517c4c9b5477aaa114338ce816c3e81/js/h5p.js#L939.

```typescript
{
  standalone: true,
  metadata?: Object,
  previousState?: Object,
  subContentId?: string,
}
```

**Comment by otacke** [comment]
@0xMurage As I explained before, it is the `H5PIntegration` object that should carry the `reportingIsEnabled` flag. It is not set by H5P Core (as we know it), but should be defined by the respective H5P integration (hence the object name) aka plugin.
Please compare the two H5P integrations for moodle as a reference:
- https://github.com/h5p/moodle-mod_hvp/blob/e0255812be2c4b74dae34650da79dc945cdb6b80/locallib.php#L79
- https://github.com/moodle/moodle/blob/7d6f282aef057fac4de4e69cff4c420f08955c51/h5p/classes/player.php#L476

**Comment by 0xMurage** [comment]
Thank you for this explanation @otacke .

But unless I misunderstand something, shouldn't the reporting be enabled if the `H5PIntegration.postUserStatistics` and `H5PIntegration.ajax.setFinished` properties are set?

<img width="873" alt="Screenshot 2025-01-03 at 15 45 57" src="https://github.com/user-attachments/assets/c383f193-4b98-4fa4-b46a-356862b06fbb" />
This is a screenshot from https://h5p.org/creating-your-own-h5p-plugin

I used these docs when documenting the [H5P standalone docs](https://github.com/tunapanda/h5p-standalone#h5p-options). 

_Support for`H5PIntegration.reportingIsEnabled` [flag](https://github.com/tunapanda/h5p-standalone/blob/3eb5591fd3ee9ed7b06023099f42adacf99a397a/src/h5p-standalone.ts#L55) has already been added to h5p standalone specifically for interactive book._

**Comment by otacke** [comment]
@0xMurage A fair share of H5P's official documentation is outdated. H5PGroup chose to use that explicit flag instead of deriving the information from other sources. You will have to ask them about the reason, however.

**Comment by csidirop** [comment]
Hi, just wanted to mention that the H5P Plugin for ILIAS got an update (see [commit 8575604](https://github.com/srsolutionsag/H5P/commit/857560454fc2e21f7f758e6e268c4a9c12c7cb31)) where they:

> - Improved the result handling so users can submit multiple results, if "solve contents only once" is not active.
> - Improved the result overview of H5P contents inside repository objects.
> - Improved the content overview of H5P repository objects.

https://github.com/srsolutionsag/H5P/blob/1d3cff9ea9c96695601c984986bc1544e3c895cd/CHANGELOG.md?plain=1#L5-L7

But it still dont fixes that issue.

**Comment by otacke** [comment]
@csidirop Mentioning this here does not make a difference. As long as the ILIAS plugin does not set the `H5PIntegration.reportingIsEnabled` flag in https://github.com/srsolutionsag/H5P/blob/release_8/classes/Integration/class.ilH5PClientDataProvider.php#L162-L207 you will not notice anything changing.

**Comment by csidirop** [comment]
Yes I know, just wanted to mention that for the sake of completeness.

**Comment by thibsy** [comment]
Hi @otacke, 

Thx a lot for your investigation.

I am reaching out on behalf of sr solutions ag (formerly known as studer & raimann ag), the developers of the ILIAS plugin.

We are wondering if there is any kind of documentation about the `H5PIntegration.reportingIsEnabled` property, as we feel uncomfortable with setting a global value of which we do not understand the functionality behind. If this property is set, what does the integration need to do? Will the AJAX endpoint be invoked as done for other content types where "results" are submitted? Can we treat these results the same? Is this flag also used in other locations or content types? If not, why is the integration required to set the flag?

Thx for your time.

Kind regards,
@thibsy

**Comment by otacke** [comment]
@thibsy No, this is not documented anywhere as far as I know of.

The integration does not _need_ to do anything. But setting this flag is supposed to signal H5P core and thus H5P content types that the integration handles xAPI (scoring) data (and offers to show reports to users in some form). That flag is in no way linked to xAPI statements (results) directly.

Some H5P libraries query that flag and change their behavior based on the setting, and that is why you may want to set it if your H5P integration supports xAPI processing/reporting.

These are the H5P libraries that I know of making use of that flag:
- _Interactive Book_ will only show the submit button if that flag is set. Otherwise, Interactive Book assumes that the platform does not support reporting and does not see the need to offer submission. As a consequence, without that flag, the `completed` xAPI statement will never be sent by Interactive Book.
- _Course Presentation_ will display a confirmation dialog before going to the summary slide when the flag is set, because going to that slide will submit scores (and usually count as an attempt).
- _Documentation Tool_ will display a submit button only if that flag is set. Otherwise, exporting to a document is the only option.
- _H5P.Question_, a library used by many question like content types, will make the "Check" button read "Submit" if that flag is set.
- _Question Set_ will make the "Finish" button read "Submit" if that flag is set.
- _Game Map_, if the flag is set, will inform the user that the score will be submitted to the platform if the user finishes a map.
